src/api.js

```diff
@@ -0,0 +1,180 @@
+  serviceName: 'events',
+  version: 'v1',
+  context: ['connection'],
+});
+
+// Returns JSON.parse(bindings) if everything goes well
+//   {"bindings" : [ 
+//     {"exchange" :  "a/b/c", "routingKey" : "a.b.c"},
+//     {"exchange" :  "x/y/z", "routingKey" : "x.y.z"},
+//   ]};
+var parseAndValidateBindings = function(bindings) {
+  return new Promise((resolve, reject) => {
+    try {
+      let json_bindings = JSON.parse(bindings);
+      if (String(Object.keys(json_bindings)) !== 'bindings') {
+        throw new Error('The json query should have only one key i.e. `bindings`.');
+      }  
+
+      // Reduce json_bindings to an array of exchanges.
+      json_bindings = json_bindings.bindings;
+      if (!Array.isArray(json_bindings)) {
+        throw new Error('Bindings must be an array of {exchange, routingKey}');
+      }
+      _.forEach(json_bindings, binding => {
+        if (!('routingKey' in binding) || !('exchange' in binding)) {
+          throw new Error('Binding must include `exchange` and `routingKey` fields');
+        }
+      });
+      resolve(json_bindings);
+    } catch (e) {
+      // A 404 code is required to send the error message without leaking internal information
+      reject({code:404, message:e.message});
+    }
+  });
+};
+
+builder.declare({
+  method: 'get',
+  route: '/connect/',
+  query: {
+    bindings: /./,
+  },
+  name: 'connect',
+  description: 'Connect to receive messages',
+  stability: APIBuilder.stability.experimental,
+  // Add input validation yml
+  title: 'Events-Api',
+}, async function(req, res) {
+  debug('hello');
+
+  // If the last event id is '-', send a 204 error blocking all reconnects.
+  // No reconnect on 204 is not yet supported on EventSource.
+  // Clients using that need to use es.close() to stop error messages.
+  if (req.headers['last-event-id']) {
+    return res.reportError(204, 'Not allowing reconnects');
+  }
+
+  let abort, headWritten;
+  const aborted = new Promise((resolve, reject) => abort = reject);
+
+  const sendEvent = (kind, data) => {
+    try {
+      var event = ['event: ' + kind,
+        'data: ' + JSON.stringify(data),
+        'id: -',
+        '\n',
+      ].join('\n');
+      
+      res.write(event);
+      debug('..sendEvent', event); 
+    } catch (err) {
+      debug('Error in sendEvent:');
+      abort(err);
+    }
+  };
+
+  try {
+
+    res.writeHead(200, {
+      'Content-Type': 'text/event-stream',
+      'Cache-Control': 'no-cache',
+    });
+    headWritten = true;
+
+    let json_bindings = await parseAndValidateBindings(req.query.bindings);
+    
+    var listener = new taskcluster.PulseListener({
+      prefetch:   5,
+      connection: this.connection,
+      maxLength:  50,
+    });
+    
+    _.forEach(json_bindings, entry => listener.bind({
+      exchange: entry.exchange,
+      routingKeyPattern: entry.routingKey,
+    }));
+    
+    listener.resume().then(
+      () => {sendEvent('ready', {});}, 
```

Comment:
Instead of sending `{}` you can make use of ES6 default function parameters and set that to `{}` if no data is specified.

```js
const sendEvent = (kind, data = {}) => {
  // ...
}

// ...

listener.resume().then(
  () => sendEvent('ready'),
```

Reply:
okay, that sounds good.